File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    int m;
    int n;

    bool findIsland(vector<vector<char>>& grid, vector<vector<bool>>& visited, int i, int j) {
        if (i < 0 || i >= m || j < 0 || j >= n) return false;
        if (visited[i][j]) return false;
        if (grid[i][j] == '0') return false;

        visited[i][j] = true;

        findIsland(grid, visited, i + 1, j);
        findIsland(grid, visited, i - 1, j);
        findIsland(grid, visited, i, j + 1);
        findIsland(grid, visited, i, j - 1);

        return true;
    }

    int numIslands(vector<vector<char>>& grid) {
        m = grid.size();
        n = grid[0].size();
        
        vector<vector<bool>> visited(m, vector<bool>(n, false));
        int count = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (!visited[i][j] && grid[i][j] == '1') {
                    bool found = findIsland(grid, visited, i, j);
                    if (found) count++;
                }
            }
        }

        return count;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def dfs(self, grid: List[List[str]], i: int, j: int):
        m, n = len(grid), len(grid[0])
        grid[i][j] = '#'
        for di, dj in ((-1, 0), (1, 0), (0, -1), (0, 1)):
            newi = i + di
            newj = j + dj
            if 0 <= newi < m and 0 <= newj < n and grid[newi][newj] == '1':
                self.dfs(grid, newi, newj)
    
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    self.dfs(grid, i, j)
                    count += 1
        return count